src/js/integrations/stalltracking.ts

```diff
@@ -0,0 +1,344 @@
+   */
+  public registerTransactionStart(
+    transaction: Transaction
+  ): (endTimestamp?: number) => StallMeasurements | null {
+    if (this._runningTransactions.has(transaction)) {
+      logger.error(
+        "[StallTracking] Tried to start stall tracking on a transaction already being tracked. Measurements might be lost."
+      );
+
+      return () => null;
+    }
+
+    this._startTracking();
+
+    if (transaction.spanRecorder) {
+      // eslint-disable-next-line @typescript-eslint/unbound-method
+      const originalAdd = transaction.spanRecorder.add;
+
+      transaction.spanRecorder.add = (span: Span): void => {
+        originalAdd.apply(transaction.spanRecorder, [span]);
+
+        // eslint-disable-next-line @typescript-eslint/unbound-method
+        const originalSpanFinish = span.finish;
+
+        span.finish = (endTimestamp?: number) => {
+          // We let the span determine its own end timestamp as well in case anything gets changed upstream
+          originalSpanFinish.apply(span, [endTimestamp]);
+
+          // The span should set a timestamp, so this would be defined.
+          if (span.endTimestamp) {
+            this._markSpanFinish(transaction, span.endTimestamp);
+          }
+        };
+      };
+    }
+
+    this._runningTransactions.add(transaction);
+    this._longestStallsByTransaction.set(transaction, 0);
+
+    const statsOnStart = this._getCurrentStats(transaction);
+
+    return (endTimestamp?: number) =>
+      this._onFinish(transaction, statsOnStart, endTimestamp);
+  }
+
+  /**
+   * Logs a transaction as finished.
+   * Stops stall tracking if no more transactions are running.
+   * @returns The stall measurements
+   */
+  private _onFinish(
+    transaction: Transaction | IdleTransaction,
+    statsOnStart: StallMeasurements,
+    passedEndTimestamp?: number
+  ): StallMeasurements | null {
+    const endTimestamp = passedEndTimestamp ?? transaction.endTimestamp;
+
+    const spans = transaction.spanRecorder
+      ? transaction.spanRecorder.spans
+      : [];
+    const finishedSpanCount = spans.reduce(
+      (count, s) => (s !== transaction && s.endTimestamp ? count + 1 : count),
+      0
+    );
+
+    const trimEnd = transaction.toContext().trimEnd;
+    const endWillBeTrimmed = trimEnd && finishedSpanCount > 0;
+
+    /*
+      This is not safe in the case that something changes upstream, but if we're planning to move this over to @sentry/javascript anyways,
+      we can have this temporarily for now.
+    */
+    const isIdleTransaction = "activities" in transaction;
+
+    let statsOnFinish: StallMeasurements | undefined;
+    if (endTimestamp && isIdleTransaction) {
+      /*
+        There is different behavior regarding child spans in a normal transaction and an idle transaction. In normal transactions,
+        the child spans that aren't finished will be dumped, while in an idle transaction they're cancelled and finished.
```

Comment:
I think this is just the current state of affairs but we should have a chat with the team and align one way or the other.

Doesn't mean this blocks the PR, but it's something we need to address.